File: 450-DeleteNodeInABst/450-DeleteNodeInABst.cpp
// Last updated: 08/05/2026, 09:29:30
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(TreeNode* &root, int &key){
        if(!root){
            return nullptr;
        }
        if(root->val>key){
            root->left =dfs(root->left, key);
        }
        else if(root->val<key){
            root->right= dfs(root->right,key);
        }
        else{
            if(!root->left){
                return root->right;
            }
            else if(!root->right){
                return root->left;
            }

            TreeNode* curr= root->right;
            while(curr->left){
                curr= curr->left;
            }
            root->val= curr->val;

            root->right= dfs(root->right, root->val);

        return root;

        }
        return root;      
    }
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root){
            return nullptr;
        }

        return dfs(root,key);
        
    }
};

File: 450-DeleteNodeInABst/450-DeleteNodeInABst.py
# Last updated: 08/05/2026, 09:29:25
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if root is None:
            return None

        if root.val<key:
            root.right= self.deleteNode(root.right, key)
        elif root.val>key:
            root.left= self.deleteNode(root.left, key)
        else:

            if root.right is None:
                return root.left
            elif root.left is None:
                return root.right
            else:

            # Two children. Inorder successor

                successor= root.right

                while successor.left:
                    successor= successor.left

                root.val= successor.val

                root.right= self.deleteNode(root.right, successor.val)

        return root
        